File: platform/apple/Sources/FipsBle/HostBleProtocol.swift
import Foundation

public enum HostBleCommand: Equatable {
    case listen(requestId: UInt64, preferredPsm: UInt16)
    case stopListening
    case startAdvertising(requestId: UInt64, bootstrap: Data)
    case stopAdvertising(requestId: UInt64)
    case startScanning(requestId: UInt64)
    case stopScanning
    case connect(requestId: UInt64, peerToken: String, psm: UInt16)
    case write(requestId: UInt64, connectionId: UInt64, bytes: Data)
    case close(connectionId: UInt64)
}

public enum HostBleEvent: Equatable {
    case listening(requestId: UInt64, psm: UInt16)
    case advertisingStarted(requestId: UInt64)
    case advertisingStopped(requestId: UInt64)
    case scanningStarted(requestId: UInt64)
    case peerDiscovered(peerToken: String, bootstrap: Data)
    case connected(
        requestId: UInt64,
        connectionId: UInt64,
        peerToken: String,
        sendSegmentMtu: UInt16,
        receiveSegmentMtu: UInt16
    )
    case incomingConnection(
        connectionId: UInt64,
        peerToken: String,
        sendSegmentMtu: UInt16,
        receiveSegmentMtu: UInt16
    )
    case bytesReceived(connectionId: UInt64, bytes: Data)
    case writeCompleted(requestId: UInt64)
    case disconnected(connectionId: UInt64, reason: String?)
    case failed(requestId: UInt64, message: String)
}

public typealias HostBleEventSink = (HostBleEvent) -> Void

File: platform/android/fips-ble/src/main/kotlin/org/fips/ble/HostBleProtocol.kt
package org.fips.ble

sealed interface HostBleCommand {
    data class Listen(val requestId: Long, val preferredPsm: Int) : HostBleCommand
    data object StopListening : HostBleCommand
    data class StartAdvertising(val requestId: Long, val bootstrap: ByteArray) : HostBleCommand
    data class StopAdvertising(val requestId: Long) : HostBleCommand
    data class StartScanning(val requestId: Long) : HostBleCommand
    data object StopScanning : HostBleCommand
    data class Connect(val requestId: Long, val peerToken: String, val psm: Int) : HostBleCommand
    data class Write(val requestId: Long, val connectionId: Long, val bytes: ByteArray) : HostBleCommand
    data class Close(val connectionId: Long) : HostBleCommand
}

sealed interface HostBleEvent {
    data class Listening(val requestId: Long, val psm: Int) : HostBleEvent
    data class AdvertisingStarted(val requestId: Long) : HostBleEvent
    data class AdvertisingStopped(val requestId: Long) : HostBleEvent
    data class ScanningStarted(val requestId: Long) : HostBleEvent
    data class PeerDiscovered(val peerToken: String, val bootstrap: ByteArray) : HostBleEvent
    data class Connected(
        val requestId: Long,
        val connectionId: Long,
        val peerToken: String,
        val sendSegmentMtu: Int,
        val receiveSegmentMtu: Int,
    ) : HostBleEvent

    data class IncomingConnection(
        val connectionId: Long,
        val peerToken: String,
        val sendSegmentMtu: Int,
        val receiveSegmentMtu: Int,
    ) : HostBleEvent

    data class BytesReceived(val connectionId: Long, val bytes: ByteArray) : HostBleEvent
    data class WriteCompleted(val requestId: Long) : HostBleEvent
    data class Disconnected(val connectionId: Long, val reason: String?) : HostBleEvent
    data class Failed(val requestId: Long, val message: String) : HostBleEvent
}

fun interface HostBleEventSink {
    fun emit(event: HostBleEvent)
}
